src/org/jgroups/protocols/raft/FileBasedLog.java

```diff
@@ -166,11 +172,42 @@ public int lastAppended() {
    }
 
    public void setSnapshot(ByteArray sn) {
-      snapshot=sn;
+      Path snapshotPath = snapshotPath();
+      try {
+         if (Files.exists(snapshotPath)) {
+            // write to temporary file first
+            Path tmp = Files.createTempFile(logDir.toPath(), null, null);
+            writeSnapshot(sn, tmp);
+            // do we need atomic move?
+            Files.move(tmp, snapshotPath, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
+         } else {
+            writeSnapshot(sn, snapshotPath);
+         }
+      } catch (IOException e) {
+         throw new RuntimeException(e);
+      }
    }
 
    public ByteArray getSnapshot() {
-      return snapshot;
+      Path snapshotPath = snapshotPath();
+      if (Files.exists(snapshotPath)) {
+         try {
+            return new ByteArray(Files.readAllBytes(snapshotPath));
+         } catch (IOException e) {
+            throw new RuntimeException(e);
+         }
+      }
+      return null;
+   }
+
+   private Path snapshotPath() {
+      return logDir.toPath().resolve(SNAPSHOT_FILE_NAME);
+   }
+
+   private static void writeSnapshot(ByteArray snapshot, Path path) throws IOException {
+      try (OutputStream os = Files.newOutputStream(path, StandardOpenOption.WRITE, StandardOpenOption.CREATE)) {
```

Comment:
If possible, would be better to use direct `ByteBuffer` to hold the `snapshot`: having big `byte[]' is dangerous both for GC (causing humoungous allocations) and for the I/O, because native I/O don't understand byte[] heap objects and is forced to copy everything into a direct ByteBuffer (or worst, by using malloc to allocate one and free when completed)

Reply:
I created #174